dot/core/service_test.go

```diff
@@ -0,0 +1,651 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package core
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/blocktree"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/keystore"
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	mocksruntime "github.com/ChainSafe/gossamer/lib/runtime/mocks"
+	rtstorage "github.com/ChainSafe/gossamer/lib/runtime/storage"
+	"github.com/ChainSafe/gossamer/lib/transaction"
+	"github.com/ChainSafe/gossamer/lib/trie"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var errTestDummyError = errors.New("test dummy error")
+var testWasmPaths []string
+
+func TestGenerateWasm(t *testing.T) {
+	wasmFilePaths, err := runtime.GenerateRuntimeWasmFile()
+	require.NoError(t, err)
+	testWasmPaths = wasmFilePaths
+}
+
+func Test_Service_StorageRoot(t *testing.T) {
+	emptyTrie := trie.NewEmptyTrie()
+	ts, err := rtstorage.NewTrieState(emptyTrie)
+	require.NoError(t, err)
+	tests := []struct {
+		name         string
+		service      *Service
+		exp          common.Hash
+		retTrieState *rtstorage.TrieState
+		retErr       error
+		expErr       error
+		expErrMsg    string
+	}{
+		{
+			name:      "nil storage state",
+			service:   &Service{},
+			expErr:    ErrNilStorageState,
+			expErrMsg: ErrNilStorageState.Error(),
+		},
+		{
+			name:      "storage trie state error",
+			service:   &Service{},
+			retErr:    errTestDummyError,
+			expErr:    errTestDummyError,
+			expErrMsg: errTestDummyError.Error(),
+		},
+		{
+			name:    "storage trie state ok",
+			service: &Service{},
+			exp: common.Hash{0x3, 0x17, 0xa, 0x2e, 0x75, 0x97, 0xb7, 0xb7, 0xe3, 0xd8, 0x4c, 0x5, 0x39, 0x1d, 0x13, 0x9a,
+				0x62, 0xb1, 0x57, 0xe7, 0x87, 0x86, 0xd8, 0xc0, 0x82, 0xf2, 0x9d, 0xcf, 0x4c, 0x11, 0x13, 0x14},
+			retTrieState: ts,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			s := tt.service
+			if tt.name != "nil storage state" {
```

Comment:
IMO this is really ugly and difficult to read.  Changing the name of the test would result in test failure? I'd consider not even making this test run in parallel in this case.  The parent `Test_Service_StorageRoot` can run in parallel yes, but I don't see the harm in running the subtests consecutively.

Reply:
@timwu20 yup I have changed it to check a parameter like quentin had suggested. Besides making the subtests no longer run in parallel, is there anything else I need to do to clean this up?

Reply:
Maybe this will remove flakiness

Reply:
I don't see how running tests in parallel or not is relevant here though?

Reply:
I discussed this with @jimjbrettj  over zoom.  With the introduction of gomock controllers and it's reliance on a locally scoped `t *testing.T` we may not be able to use the IDE code generation for tests if we want the subtests to run in parallel.  

Yes detecting race conditions would be better if each subtest is run in parallel, but the tradeoff is we can't use the code generated tests with predefined state if we are using gomock controllers.  In the past I just created my own mock implementations of interfaces that did not rely on `t *testing.T`.  Rather, just emit testing errors when assertions fail or through inspection of the mock implementation.

In this case specifically, it may just make more sense to avoid using the slice of anonymous test structs and just adopt the explicit way of writing out each subtest.  We should avoid repeating code over and over again, which I have discussed, and it will allow us to keep using gomock controllers and run the subtests in parallel.

Reply:
> we may not be able to use the IDE code generation for tests if we want the subtests to run in parallel.

It's not necessarily about running in parallel. If one subtest fails because of a mock assertion failing, it might fail the parent test since it would take `t` from the parent test, but would run within the subtest. I guess it's *alriiight* to share a mock controller  at the parent test, although it's not ideal either. For test generation, you could use Github copilot it does help quite a bit to have a good middleground between manual and automated test writing.

> Rather, just emit testing errors when assertions fail or through inspection of the mock implementation.

That's the same problem though, since it would fail the parent test if one of the subtest fails for example. This is the same as passing `t` once to a subtest-local mock controller.

> In this case specifically, it may just make more sense to avoid using the slice of anonymous test structs and just adopt the explicit way of writing out each subtest. 

Agree depending on the case. If mocks configuration is quite different from one test case to another (i.e. a mock depending on another mock, for certain test cases), then yes dumb subtests without for loop / test case slices is best. Otherwise, configure your mock within the subtest. I wrote [this comment](https://github.com/ChainSafe/gossamer/pull/2239#discussion_r803765992) which sums it up from my perspective.

Reply:
Created https://github.com/golang/mock/issues/625 if that interests anyone 😄
But yes, ultimately, controller construction within the subtest is pretty good for local scoping / readability / parallel tests.

Reply:
Added some code style section for mocks configuration in subtests at https://github.com/ChainSafe/gossamer/pull/2311

Reply:
Tim found a new way to do mocks with a test case slice/map + for loop and subtests. However I don't think it works when a mock needs to return a mock for example.

1. For each mock, add a mock constructor field in your test case with signature `func(ctrl *gomock.Controller) YourInterface`
2. Configure your mock in the test case by defining the function body
3. Call that function from your test case in the subtest, passing it the subtest's Gomock controller as argument

Again, if you need to i.e. return a mock from another mock, this method doesn't seem to work and in this case you need to use one of the two options mentioned in the codestyle document.